internal/backend/local/test.go

```diff
@@ -345,930 +339,47 @@ func (runner *TestFileRunner) walkGraph(g *terraform.Graph, file *moduletest.Fil
-				// execute tests. Instead, we mark all remaining run blocks as
-				// skipped, print the status, and move on.
-				run.Status = moduletest.Skip
-				runner.Suite.View.Run(run, file, moduletest.Complete, 0)
-				return
-			}
-			if runner.Suite.Stopped {
-				// Then the test was requested to be stopped, so we just mark each
-				// following test as skipped, print the status, and move on.
-				run.Status = moduletest.Skip
-				runner.Suite.View.Run(run, file, moduletest.Complete, 0)
-				return
-			}
-
-			// TODO: The execution of a NodeTestRun is currently split between
-			// its Execute method and the continuation of the walk callback.
-			// Eventually, we should move all the logic related to a test run into
-			// its Execute method, effectively ensuring that the Execute method is
-			// enough to execute a test run in the graph.
-			diags = v.Execute(runner.EvalContext)
-			if diags.HasErrors() {
-				return diags
-			}
-
-			startTime := time.Now().UTC()
-			runner.run(run, file, startTime)
-			runner.Suite.View.Run(run, file, moduletest.Complete, 0)
-			file.UpdateStatus(run.Status)
-		case graph.GraphNodeExecutable:
-			diags = v.Execute(runner.EvalContext)
-			return diags
-		default:
-			// If the vertex isn't a test run or executable, we'll just skip it.
-			return
+		// Bind the terraform context options to the node
+		if v, ok := v.(graph.BindContextOpts); ok {
+			v.BindContextOpts(runner.Suite.Opts)
 		}
-		return
-	}
-
-	return g.AcyclicGraph.Walk(walkFn)
-}
 
-func (runner *TestFileRunner) run(run *moduletest.Run, file *moduletest.File, startTime time.Time) {
-	log.Printf("[TRACE] TestFileRunner: executing run block %s/%s", file.Name, run.Name)
-	defer func() {
-		// If we got far enough to actually execute the run then we'll give
-		// the view some additional metadata about the execution.
-		run.ExecutionMeta = &moduletest.RunExecutionMeta{
-			Start:    startTime,
-			Duration: time.Since(startTime),
-		}
-
-	}()
-
-	key := run.GetStateKey()
-	if run.Config.ConfigUnderTest != nil {
-		if key == moduletest.MainStateIdentifier {
-			// This is bad. It means somehow the module we're loading has
-			// the same key as main state and we're about to corrupt things.
-
-			run.Diagnostics = run.Diagnostics.Append(&hcl.Diagnostic{
-				Severity: hcl.DiagError,
-				Summary:  "Invalid module source",
-				Detail:   fmt.Sprintf("The source for the selected module evaluated to %s which should not be possible. This is a bug in Terraform - please report it!", key),
-				Subject:  run.Config.Module.DeclRange.Ptr(),
-			})
-
-			run.Status = moduletest.Error
-			file.UpdateStatus(moduletest.Error)
+		// handle test run node separately
+		if runNode, ok := v.(*graph.NodeTestRun); ok {
+			file := runNode.File()
+			diags = runNode.Execute(runner.EvalContext)
+			runner.EvalContext.Renderer().Run(runNode.Run(), file, moduletest.Complete, 0)
+			file.UpdateStatus(runNode.Run().Status)
```

Comment:
Could you do this in the Execute function for the NodeTestRun? Then we don't need this extra if statement here with the special handling.